feature(application): adds services to Application and adds elgg()

mrclay opened the pull request:
elgg() provides access to the booted Application instance which will provide access to future public API service objects (instead of global functions).

As a first step, this adds only the “config” service, which is needed to use the method getVolatile() before site boot.

For #8238

mrclay commented on engine/classes/Elgg/Application.php:
```diff
@@ -27,6 +32,18 @@ class Application {
+		'events',
+		'hooks',
+	];
+
```
These were just a few starters. So many decisions to make...

ewinslow commented on engine/classes/Elgg/Application.php:
```diff
@@ -183,6 +175,8 @@ public function loadCore() {
 				}
 			}
 
+			elgg($this);
```
this line is rather unclear... what in the world does this do?

mrclay replied:
Why it Elgg's this thing obviously! Will add comment. :)

juho-jaakkola replied:
:)

ewinslow commented on engine/tests/phpunit/Elgg/ApplicationTest.php:
```diff
@@ -0,0 +1,49 @@
+		$this->assertSame($services->hooks, $app->hooks);
+	}
+
+	function testLoadSettings() {
```
test names should also include expected result.

ewinslow commented:
I like the direction. Wondering if we should skip to php-di instead and map from interface names to instances rather than string keys to instances?

ewinslow commented:
I.e.:

```php

use Elgg\Context;

elgg()->get(Context::class)->push('foo');
```

as opposed to:

```php
elgg()->context->push('foo');
```

The latter is certainly more terse but it doesn't have the nice, automatic namespacing benefit that comes with using classes/interfaces as keys.

ewinslow commented:
Obviously in older version of PHP that would have to be `elgg()->get('Elgg\Context')->push('foo');`

mrclay commented:
What's the "namespacing benefit"? The huge advantage of faux properties/methods is IDE static analysis/autocomplete. Really we could support both syntaxes easily like we'd planned to before.

mrclay commented:
Having explicit core class names in 3rd party plugin code seems like a downside to me, considering I can jump straight to the method definition in the IDE. If plugins don't bind to class names, that seems more flexible for core.

mrclay commented:
Latest push adds "session" and "site" (both already accessible from current API) and uses an Application static prop to simplify elgg().

ewinslow commented:
> Having explicit core class names in 3rd party plugin code seems like a downside to me

This is not something we'll be able to avoid when we start supporting ctor/method injection. But that still comes with ide support also, so I'm fine with this for now.

ewinslow commented:
If we do start leaking core class names into plugin code, we should strive to make this only be interface names, rather than concrete classes...

Srokap commented:
> If we do start leaking core class names into plugin code, we should strive to make this only be interface names, rather than concrete classes...

Very much agreed.

mrclay commented:
While the Application class is a step forward, I'm less excited about the process of slowly opening up existing services, making interfaces, etc. As I've blogged about before, this is a huge undertaking that is a big distraction away from more high value work we could be doing on the front end.

I'd suggest we reserve elgg() for newly created APIs rather than making duplicates of existing APIs.

ewinslow commented:
Speaking of front end, I have some new ideas there I'm really excited about... Will post elsewhere...

juho-jaakkola commented:
> I'd suggest we reserve elgg() for newly created APIs rather than making duplicates of existing APIs.

Not counting the ones already added to this PR?

mrclay commented:
I think it would be messy to make an arbitrary chunk of our API accessible two ways. It widens the surface area we have to support and document, and these services weren't really designed for dev usability, just quickly copied to classes.

So yes I'd be in favor of removing the pre-existing services I've exposed here.

mrclay committed:
feature(application): adds elgg() and makes Application a service provider

elgg() provides access to the booted Application instance which will
provide access to future public API service objects (instead of global
functions).

As a first step, this adds only the “config” service, which is needed to
use the method getVolatile() before site boot.

mrclay commented:
I've rolled this back a bit to just introduce elgg() and provide the "config" service, which was already exposed anyway in the master branch. It doesn't add docs as there's no reason yet for devs to use it.

ewinslow commented:
Sgtm

mrclay commented:
FYI, in #8288 I've changed the "config" service to be exposed as an interface, and documented services and service creation.
